pennylane/hf/tapering.py

```diff
@@ -464,3 +467,106 @@ def transform_hamiltonian(h, generators, paulix_ops, paulix_sector):
+    r"""Get the optimal sector which contains the ground state.
+
+    To obtain the optimal sector one can brute force through all the permutation or by utilize the following
+    relation between the Pauli-Z qubit operator and the occupation number under Jordan-Wigner transform.
+
+    .. math::
+
+        \sigma_z^{i} = a_{i}^{\dagger}a_{i} - 1
+
+    This relation allows us to figure out whether the orbital is occupied or unoccupied in a given symmetry sector,
+    to build the correct eigensector.
+
+    Args:
+        qubit_op (pennylane.Hamiltonian): Hamiltonian for which symmetries are being generated for performing tapering
+        generators (list[pennylane.Hamiltonian]): list of generators of symmetries, taus, for the Hamiltonian
+        paulix_ops (list[pennylane.operation.Observable]):  list of single-qubit Pauli X operators
+        brute_force (bool): determines whether to use brute-force strategy to pick the correct sector or not
+        num_electrons (int): If `brute_force = True`, user must provide the number of active electrons in
+                              the system for generating the Hartree-Fock bitstring
+
+    Returns:
+        tuple (list[int], float):
+
+            * list[int]: eigenvalues corresponding to the optimal sector which contains the ground state
+            * float: energy of the computed optimal sector
+
+    .. code-block:: python
+
+        >>> symbols = ['H', 'H']
+        >>> coordinates = np.array([0., 0., -0.66140414, 0., 0., 0.66140414]))
+        >>> mol = qml.hf.Molecule(symbols, coordinates)
+        >>> H, qubits = qml.hf.generate_hamiltonian(mol)(), 4
+        >>> generators, paulix_ops = generate_symmetries(H, qubits)
+        >>> find_optimal_sector(H, generators, paulix_ops, False, 2)
+          ([1, -1, -1], -1.1372701746609024)
+
+    """
+
+    if not brute_force:
+
+        if num_electrons < 1:
+            raise ValueError(
+                f"Brute force search is disabled;"
+                f"the number of electrons must be provided and should be greater than zero;"
+                f"got 'electrons'={num_electrons}"
+            )
+
+        num_orbitals = len(qubit_op.wires)
+
+        if num_electrons > num_orbitals:
+            raise ValueError(
+                f"Number of active orbitals cannot be smaller than number of active electrons;"
+                f" got 'orbitals'={num_orbitals} < 'electrons'={num_electrons}."
+            )
+
+        hf_str = np.where(np.arange(num_orbitals) < num_electrons, 1, 0)
+        perm = []
+
+        for op in generators:
+            symmstr = np.zeros(len(qubit_op.wires), dtype=int)
+            for wire in op.wires:
+                symmstr[wire] = 1
+            coeff = -1 if numpy.logical_xor.reduce(numpy.logical_and(symmstr, hf_str)) else 1
+            perm.append(coeff)
+
+        sector = transform_hamiltonian(qubit_op, generators, paulix_ops, perm)
+        energy = _sector_energy(sector)
```

Comment:
Not sure we should return the energy here since diagonalizing the Hamiltonian is very expensive